python/ray/_raylet.pyx

```diff
@@ -2228,7 +2228,26 @@ cdef execute_task_with_cancellation_handler(
 
         if omp_num_threads_overriden:
             # Reset the OMP_NUM_THREADS environ if it was set.
-            os.environ.pop("OMP_NUM_THREADS", None)
+            try:
+                os.environ.pop("OMP_NUM_THREADS", None)
+            except KeyError as e:
+                # os.environ is known to have undefined behavior if multiple 
+                # threads is trying to modify the mapping simultaneously. 
+                # Here specifically, race condition could happen when two threads 
+                # poping the environ in the same time, where one of them deleting 
+                # the key successfully, but the other met KeyError. If we met this 
+                # issue, we just ensure the the current os.environ do not contain 
+                # this env var and move on. 
+                # Related issue: https://github.com/python/cpython/issues/120513
+                logger.error(
+                    "KeyError occurred when popping OMP_NUM_THREADS from "
+                    "os.environ, a possible race condition might happened. "
+                    "Checking if the env var is already cleaned up and move on."
+                )
+                if os.environ.get("OMP_NUM_THREADS", None) is not None:
```

Comment:
When will this happen? Only condition on top of my mind is it got added back by another thread, but that thread will eventually pop it out.

Reply:
The only way to prevent this from happening is to introduce a lock on any access to os.environ. If we are not doing so then there is always going to be a small chance that this could happen. It is hard for me to come up with an exhausted list. But adding a logging here will make sure this is easier to know in the future.